contracts/tokenbridge/ethereum/L1AtomicTokenBridgeCreator.sol

```diff
@@ -0,0 +1,506 @@
+
+/**
+ * @title Layer1 token bridge creator
+ * @notice This contract is used to deploy token bridge on custom L2 chains.
+ * @dev Throughout the contract terms L1 and L2 are used, but those can be considered as base (N) chain and child (N+1) chain
+ */
+contract L1AtomicTokenBridgeCreator is Initializable, OwnableUpgradeable {
+    error L1AtomicTokenBridgeCreator_OnlyRollupOwner();
+    error L1AtomicTokenBridgeCreator_InvalidRouterAddr();
+    error L1AtomicTokenBridgeCreator_TemplatesNotSet();
+
+    event OrbitTokenBridgeCreated(
+        address indexed inbox,
+        address indexed owner,
+        address router,
+        address standardGateway,
+        address customGateway,
+        address wethGateway,
+        address proxyAdmin
+    );
+    event OrbitTokenBridgeTemplatesUpdated();
+    event NonCanonicalRouterSet(address indexed inbox, address indexed router);
+
+    // non-canonical router registry
+    mapping(address => address) public inboxToNonCanonicalRouter;
+
+    // Hard-code gas to make sure gas limit is big enough for L2 factory deployment to succeed.
+    // If retryable would've reverted due to too low gas limit, nonce 0 would be burned and
+    // canonical address for L2 factory would've been unobtainable
+    uint256 public gasLimitForL2FactoryDeployment;
+
+    // contract which creates retryables for deploying L2 side of token bridge
+    L1TokenBridgeRetryableSender public retryableSender;
+
+    // L1 logic contracts shared by all token bridges
+    L1GatewayRouter public routerTemplate;
+    L1ERC20Gateway public standardGatewayTemplate;
+    L1CustomGateway public customGatewayTemplate;
+    L1WethGateway public wethGatewayTemplate;
+
+    // L2 contracts deployed to L1 as bytecode placeholders
+    address public l2TokenBridgeFactoryTemplate;
+    address public l2RouterTemplate;
+    address public l2StandardGatewayTemplate;
+    address public l2CustomGatewayTemplate;
+    address public l2WethGatewayTemplate;
+    address public l2WethTemplate;
+
+    // WETH address on L1
+    address public l1Weth;
+
+    // immutable canonical addresses for L2 contracts
+    // other canonical addresses (dependent on L2 template implementations) can be fetched through `getCanonicalL2***Address` functions
+    address public canonicalL2FactoryAddress;
+    address public canonicalL2ProxyAdminAddress;
+    address public canonicalL2BeaconProxyFactoryAddress;
+
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize() public initializer {
+        __Ownable_init();
+
+        // deploy retryableSender only once - its address will be part of salt for L2 contracts
+        if (address(retryableSender) == address(0)) {
+            retryableSender = L1TokenBridgeRetryableSender(
+                address(
+                    new TransparentUpgradeableProxy(
+                        address(new L1TokenBridgeRetryableSender()),
```

Comment:
In later PR I removed L1TokenBridgeRetryableSender from here, instead it is being passed in (due to contract size limits). I will update deployment script so that it uses deterministic CREATE2 factory to deploy both `L1AtomicTokenBridgeCreator` and `L1TokenBridgeRetryableSender` so they have same address on any basechain